### PR TITLE
Support complex type 'compare' in 3-arg min_by and max_by Presto aggregates

### LINKED ISSUE #7469: Support complex type for the compare param for min_by/max_by
### Description

Presto's min_by/max_by support complex type for the compare param(second param): 

```
presto:sf1> SELECT min_by(c0, c1, c2)
         -> FROM (
         ->   VALUES
         ->     ('z', ARRAY[1, 2], 3),
         ->     ('a', ARRAY[2, 3], 3),
         ->     ('x', ARRAY[0, 3], 3),
         ->     ('b', ARRAY[0, 2], 3)
         -> ) AS tmp (c0, c1, c2);
   _col0
-----------
 [b, x, z]
(1 row)
```

While Velox does not support it, Velox currently only support primitive type for the compare param:
https://github.com/facebookincubator/velox/blob/42b604b6eed9b7d86958bc170b9f32be4018ccf8/velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp#L526-L536

### PR DESCRIPTION
**Pull request by f0rest9999**
Presto's min_by and max_by with 3 args support complex type for the compare param, such as array.

Fixes #7469

### CODE REVIEW

**Comment by f0rest9999** [comment]
@mbasmanova Could you please review this PR?

**Comment by f0rest9999** [comment]
Could someone please review this PR? Thanks！ @mbasmanova @kagamiori @kgpai

**Comment by kgpai** [comment]
Thank you for this @f0rest9999 ! This PR will take me a bit of time, please give me a couple of days to review.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 1]
```diff
@@ -2201,5 +2201,142 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
   }
 }
 
+TEST_F(MinMaxByComplexTypes, arrayTypeCompare) {
+  // The data consists of five rows, and the comparison key is arrayType.
+  // Each array contains two elements.
+  auto data = makeRowVector({
```
Comment (kgpai) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2207]: Can you add a test with stringview types (both as value and comparison ) ?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 2]
```diff
@@ -2201,5 +2201,142 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
...
+  });
+
+  expected = makeRowVector(
+      {makeFlatVector<int16_t>({1, 2}),
+       makeArrayVector<int64_t>({{8, 4}, {3, 5}})});
+
+  testAggregations({data}, {"c0"}, {"max_by(c1, c2, 2)"}, {expected});
+}
+
+TEST_F(MinMaxByComplexTypes, mapTypeCompare) {
```
Comment (kgpai) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2233]: Can you add a test with nulls in map's/arrays (and null arrays) and also validate that behavior with Presto ?

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 3]
```diff
@@ -2201,5 +2201,142 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
...
+               makeFlatVector<int64_t>({1, 2, 3, 6, 7, 8}),
+               makeFlatVector<int64_t>({10, 20, 30, 60, 70, 80})))});
+
+  testAggregations({data}, {"c0"}, {"min_by(c1, c2, 1)"}, {expected});
+}
+
+TEST_F(MinMaxByComplexTypes, rowTypeCompare) {
+  // The data consists of four rows, and the comparison key is RowType and
+  // the value key are ArrayType. Each comparison key has two attributes.
+  auto data = makeRowVector({
```
Comment (kgpai) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2298]: Curious can you add some test for : 
1. nulls in Rows 
2. Complex types in Rows

Also if possible validate this behavior with Presto.

**Review by kgpai — COMMENTED** [review]
Thanks for this change !

**Comment by f0rest9999** [comment]
Thanks for the review, I'll take the time to think about and fix.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp (hunk 2 above)
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2233]: Thanks for this. I found something when validating the behavior with Presto.

1. For Array type, Presto supports array without nulls.

Without nulls.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, ARRAY[10, 30]),
     ->     (2, 43.0, ARRAY[20, 10]),
     ->     (3, 53.0, ARRAY[50, 40])
     ->   ) AS t (id, col1, col2)
     -> ) SELECT min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [43.0, 43.0] 
```

With nulls.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, ARRAY[10, NULL, 30]),
     ->     (2, 43.0, ARRAY[20, NULL, 10]),
     ->     (3, 53.0, ARRAY[NULL, 50, 40])
     ->   ) AS t (id, col1, col2)
     -> ) SELECT min_by(col1, col2, 2)
     -> FROM dataset;
Query 20240321_092745_00006_j8s5b, FAILED, 1 node
Splits: 6 total, 0 done (0.00%)
[Latency: client-side: 107ms, server-side: 83ms] [0 rows, 0B] [0 rows/s, 0B/s]

Query 20240321_092745_00006_j8s5b failed: ARRAY comparison not supported for arrays with null elements
```
2. Presto not support MapType as key(also test in demo ut).
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 43.0, MAP(ARRAY['a', 'b'], ARRAY[10, 11])),
     ->     (2, 43.0, MAP(ARRAY['c', 'd'], ARRAY[11, 10])),
     ->     (3, 53.0, MAP(ARRAY['e', 'f'], ARRAY[50, 11]))
     ->   ) AS t (id, col1, col2)
     -> ) SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
Query 20240321_093011_00007_j8s5b failed: line 8:3: 
Unexpected parameters (decimal(3,1), map(varchar(1),integer), integer) for function min_by. Expected: min_by(V, K) K:orderable, V, min_by(V, K, bigint) V, K:orderable
```
3. Presto support RowType as key, but not support RowType col **which participate in comparison** contains null
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, 'A', TRUE)),
     ->     (2, 43.0, ROW(20, 'B', FALSE)),
     ->     (3, 53.0, ROW(50, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [42.0, 43.0] 
(1 row)
```

The second col does not participate in comparison.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, NULL, TRUE)),
     ->     (2, 43.0, ROW(20, 'B', FALSE)),
     ->     (3, 53.0, ROW(50, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;
    _col0     
--------------
 [42.0, 43.0] 
(1 row)
```

The second col participate in comparison.
```
presto> WITH dataset AS (
     ->   SELECT * FROM (VALUES
     ->     (1, 42.0, ROW(10, NULL, TRUE)),
     ->     (2, 43.0, ROW(10, 'B', FALSE)),
     ->     (3, 53.0, ROW(10, 'C', TRUE))
     ->   ) AS t (id, col1, col2)
     -> ) 
     -> SELECT
     ->   min_by(col1, col2, 2)
     -> FROM dataset;

Query 20240321_094004_00016_j8s5b, FAILED, 1 node
Splits: 6 total, 0 done (0.00%)
[Latency: client-side: 101ms, server-side: 72ms] [0 rows, 0B] [0 rows/s, 0B/s]

Query 20240321_094004_00016_j8s5b failed: ROW comparison not supported for fields with null elements
```

Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2233]: So I will look at how to validate the behavior with Presto.

remove mapType in this PR
Throw an exception when the key col which participate in comparison contains null.
How do you think?

**Comment by mbasmanova** [comment]
> Throw an exception when the key col which participate in comparison contains null.

This is a bit tricky. See kNullAsIndeterminate in CompareFlags.h and #7703

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 4]
```diff
@@ -542,8 +542,492 @@ struct ComplexTypeExtractor {
   }
 };
 
+template <typename V, typename Compare>
+struct MinMaxByNComplexCompareTypeAccumulator {
```
[the remark this answers is not in the document]
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:546]: Currently

- Both C and V are simple: MinMaxByNAccumulator
- At least one of C or V are StringView: MinMaxByNStringViewAccumulator
- Value is complex: MinMaxByNComplexTypeAccumulator

While this PR add three case and accumulator

- C is complex and V is simple: MinMaxByNComplexCompareTypeAccumulator
- C is complex and V is StringView: MinMaxByNStringViewValueTypeComplexCompareTypeAccumulator
- Both C and V are complex:MinMaxByNBothComplexTypeAccumulator

I think the benefits of this approach are that the previous accumulator is not modified, and the logic inside the accumulator is clearer.The downside is that the code is not reusable enough.

And I try to follow your suggestion to modify: Merge three accumulators that I add to MinMaxByNComplexTypeAccumulator, corresponds to the case that at least one of C or V are complex.
But this brings a problem: the method in this accumulator looks complicated.
For example, compareAndAdd method will change to this. It seems harder to read. Also other methods will change like this.
```
void compareAndAdd(
      DecodedVector& decodedCompare,
      vector_size_t compareIndex,
      DecodedVector& decodedValue,
      vector_size_t valueIndex,
      Compare& comparator) {
    C comparison;
    V value;
    if (std::is_same_v<C, ComplexType>) {
      comparison = writeComplexCompare(decodedCompare, compareIndex);
      if (std::is_same_v<V, ComplexType>) {
        value = writeComplexValue(decodedValue, valueIndex);
      } else {
        if (std::is_same_v<V, StringView>) {
          value = writeStringValue(value);
        } else {
          value = optionalValue(decodedValue, valueIndex);
        }
      }
    } else {
      comparison = decodedCompare.valueAt<C>(compareIndex);
      value = writeComplexValue(decodedValue, valueIndex);
    }

    // pseudo-code
    addToAccumulator(comparison, value)...
  }
```
So which one you prefer and any advice on that above please? Thanks!

**Comment by f0rest9999** [comment]
> nit: This class is very similar to MinMaxByNComplexTypeAccumulator save only difference is that here the compare is comples and there the value is complex, I am curious if there is a way to merge both into 1 (without adding too much templating so its hard to follow).

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 5]
```diff
@@ -2201,5 +2201,143 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
...
+
+  expected = makeRowVector(
+      {makeFlatVector<int16_t>({1, 2}),
+       makeArrayVector<int64_t>({{8, 4}, {3, 5}})});
+
+  testAggregations({data}, {"c0"}, {"max_by(c1, c2, 2)"}, {expected});
+
+  // Test StringView type (both as value and comparison).
+  data = makeRowVector({
+      makeFlatVector<StringView>({"1"_sv, "2"_sv, "3"_sv, "4"_sv}),
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2234]: This test uses only inlined strings, hence, covers only easy code path. It needs to use longer strings to cover "interesting" code paths.

Also, consider using `makeFlatVector<std::string>` for readabiity.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 6]
```diff
@@ -2201,5 +2201,143 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
   }
 }
 
+TEST_F(MinMaxByComplexTypes, arrayTypeCompare) {
+  // The data consists of five rows, and the comparison key is arrayType.
+  // Each array contains two elements.
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>({8, 3, 4, 5, 6}),
+      makeArrayVector<int64_t>({{3, 4}, {2, 4}, {2, 3}, {0, 3}, {0, 2}}),
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2209]: Let's also tests arrays with nulls, both cases where query succeeds and  when it fails.

File: velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp [hunk 7]
```diff
@@ -2201,5 +2201,143 @@ TEST_F(MinMaxByNTest, incrementalWindow) {
...
+               {3, 4, 5},
+               {6, 7},
+               {8, 9, std::nullopt},
+           }),
+           makeNullableFlatVector<int32_t>({3, std::nullopt, 1, 4})}),
+  });
+
+  expected = makeRowVector(
+      {makeFlatVector<int16_t>({1, 2}),
+       makeArrayVector(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp:2334]: Consider using makeNestedArrayVectorFromJson

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 8]
```diff
@@ -556,6 +1028,7 @@ struct Less {
 
 template <typename V, typename C>
 struct Greater {
+  explicit Greater(TypePtr& /* type */){};
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:1031]: Same question as above.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 9]
```diff
@@ -542,8 +542,480 @@ struct ComplexTypeExtractor {
...
+  void checkAndSetN(DecodedVector& decodedN, vector_size_t row) {
+    return base.checkAndSetN(decodedN, row);
+  }
+
+  void compareAndAdd(
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex,
+      std::optional<V> value,
+      Compare& comparator) {
+    auto comparePosition = writeComplex(decodedCompare, compareIndex);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:580]: This code should not run unconditionally. It should be evaluated only if we are going to store the value in the accumulator.

**Comment by f0rest9999** [comment]
@mbasmanova Already fixed.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 10]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+    return base.checkAndSetN(decodedN, row);
+  }
+
+  void compareAndAdd(
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex,
+      std::optional<V> value,
+      Compare& comparator) {
+    if (base.heapValues.size() < base.n) {
+      auto comparePosition = writeComplex(decodedCompare, compareIndex);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:581]: Let's think about how to extend tests to capture the bug I pointed out earlier. Let's also check other compareAndAdd for similar issues.

**Review by mbasmanova — COMMENTED** [review]
Flushing pending comments.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 11]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+        auto comparePosition = writeComplex(decodedCompare, compareIndex);
+        std::pop_heap(
+            base.heapValues.begin(), base.heapValues.end(), comparator);
+        base.heapValues.pop_back();
+        addToAccumulator(comparePosition, value, comparator);
+      }
+    }
+  }
+
+  /// Extract all values from 'heapValues' into 'rawValues' and 'rawValueNulls'
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:595]: This comment doesn't match the code. There are no rawValues and rawValueNulls in the signature or implementation.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 12]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+  /// buffers. The heap remains unchanged after the call.
+  void extractValues(
+      FlatVector<V>& values,
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:601]: Looks like this code repeats a lot. Perhaps, it can be refactoring into a helper function that takes 'extractValue' lambda.

```
extractValues(base.heapValues, [&] (const auto& pair, auto i) {
   extractValue(pair, values, offset + i);
}
```

Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:601]: Ditto extractPairs

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 10 above)
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:581]: > This code should not run unconditionally. It should be evaluated only if we are going to store the value in the accumulator.

You mean this above? I think this is not a bug, but to avoid unnecessary operations? (Maybe the comment I quote is not you specified^^)

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 13]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+    if (base.heapValues.size() < base.n) {
+      auto comparePosition = writeComplex(decodedCompare, compareIndex);
+      addToAccumulator(comparePosition, value, comparator);
+    } else {
+      const auto& topPair = base.heapValues.front();
+      if (comparator.compare(topPair, decodedCompare, compareIndex)) {
+        auto comparePosition = writeComplex(decodedCompare, compareIndex);
+        std::pop_heap(
+            base.heapValues.begin(), base.heapValues.end(), comparator);
+        base.heapValues.pop_back();
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:589]: When we remove an entry, should we also free memory used by that entry? If we don't, then we'll keep a lot of memory allocated unnecessarily. Let's make sure we use only needed amount of memory when processing both 'compare' values increasing (every new row causes a removal of existing vlue) and decreasing (new rows should be processed very fast without any memory allocation or deallocation).

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 10 above)
Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:581]: @f0rest9999 Indeed, this issue doesn't affect correctness, but affects performance and in extreme cases may cause query failure due to OOM. We need to ensure both correct and efficient implementation.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 14]
```diff
@@ -542,8 +542,480 @@ struct ComplexTypeExtractor {
...
+      if (comparator.compare(topPair, decodedCompare, compareIndex)) {
+        std::pop_heap(
+            base.heapValues.begin(), base.heapValues.end(), comparator);
+        base.heapValues.pop_back();
+        addToAccumulator(comparePosition, value, comparator);
+      }
+    }
+  }
+
+  /// Extract all values from 'heapValues' into 'rawValues' and 'rawValueNulls'
```
[the remark this answers is not in the document]
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:595]: OK, also found in previous code. I will fix all formatting in this cpp.

**Review by mbasmanova — COMMENTED** [review]
@f0rest9999 I wonder if it makes sense to rewrite these aggregate functions using Simple Function API. This may simplify the code and make it easier to read and extend.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 15]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+      const std::optional<V>& value,
+      Compare& comparator) {
+    base.heapValues.push_back(
+        std::make_pair(comparePosition, writeString(value)));
+    std::push_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+  }
+
+  void freePair(typename BaseType::Heap::const_reference pair) {
+    valueSet.free(pair.first.header);
+    valueSet.free(*pair.second);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:807]: How this work when pair.second is an std::nullopt?

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 16]
```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
...
+      return std::nullopt;
+    }
+    HashStringAllocator::Position position;
+    valueSet.write(*decoded.base(), decoded.index(index), position);
+    return position;
+  }
+
+  void freePair(typename BaseType::Heap::const_reference pair) {
+    valueSet.free(pair.first.header);
+    valueSet.free(pair.second->header);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:975]: What happens when pair.second is an std::nullopt?

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 17]
```diff
@@ -566,6 +1044,91 @@ struct Greater {
   }
 };
 
+template <typename V>
+struct ComplexLess {
+  TypePtr compareType;
+  explicit ComplexLess(TypePtr& type) : compareType(type){};
+  using Pair = std::pair<HashStringAllocator::Position, std::optional<V>>;
+  static CompareFlags kCompareFlags;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:1052]: Are we using default values here? Is this working correctly?

```
NullHandlingMode nullHandlingMode = NullHandlingMode::kNullAsValue;
```

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 18]
```diff
@@ -566,6 +1044,91 @@ struct Greater {
...
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex) {
+    VELOX_CHECK_NOT_NULL(topPair.first.header);
+    ByteInputStream stream =
+        HashStringAllocator::prepareRead(topPair.first.header);
+    // Returns > 0 if 'topPair' greater than 'vector' at 'index', means it needs
+    // to be pushed into the heap.
+    std::optional<int32_t> res = exec::ContainerRowSerde::compareWithNulls(
+        stream, decodedCompare, compareIndex, kCompareFlags);
+    return res.value() > 0;
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:1078]: What happens when res is an std::nullopt?

**Review by mbasmanova — COMMENTED** [review]
@f0rest9999 Would you run Aggregation Fuzzer using Presto as source of truth with --only min_by,max_by for a couple of hours to make sure there are no failures.

**Comment by f0rest9999** [comment]
That's a good suggestion. Could you open an issue and see if anyone is interested? While I probably haven't had a lot of time to look into how to implement this recently, and I prefer to get this non-simple Function API PR done first to support complex types. Is that all right? 

And I'll spend some time looking at how to test "Aggregation Fuzzer using Presto". Thanks.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 18 above)
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:1078]: image
In our case, NullHandlingMode == kNullAsIndeterminate & equalsOnly == false, it won't return std::nullopt, while it will throw an exception.

**Comment by f0rest9999** [comment]
Have you run Aggregation Fuzzer using Presto as source before? In the process of running, I found some problems. 
@mbasmanova @kgpai 

When I run the **arbitrary** functions for test (from https://facebookincubator.github.io/velox/develop/testing/fuzzer.html) , an error will be reported.

```
./velox_aggregation_fuzzer_test \
                --seed ${RANDOM} \
                --only "arbitrary" \
                --duration_sec 3600 \
                --logtostderr=1 \
                --minloglevel=0 \
                --repro_persist_path=/tmp/aggregate_fuzzer_repro \
                --enable_sorted_aggregations=true \
                --presto_url=http://127.0.0.1:8080 
I20240327 14:04:16.002938 3310529 HiveConnector.cpp:69] Hive connector test-hive created with maximum of 20000 cached file handles.
I20240327 14:04:16.003876 3310529 AggregationFuzzerBase.cpp:576] Total functions: 1 (1 signatures)
I20240327 14:04:16.004051 3310529 AggregationFuzzerBase.cpp:580] Functions with at least one supported signature: 1 (100.00%)
I20240327 14:04:16.004269 3310529 AggregationFuzzerBase.cpp:586] Functions with no supported signature: 0 (0.00%)
I20240327 14:04:16.004276 3310529 AggregationFuzzerBase.cpp:589] Supported function signatures: 1 (100.00%)
I20240327 14:04:16.004280 3310529 AggregationFuzzerBase.cpp:595] Unsupported function signatures: 0 (0.00%)
I20240327 14:04:16.004335 3310529 AggregationFuzzer.cpp:325] ==============================> Started iteration 0 (seed: 7063)
I20240327 14:04:16.020104 3310529 AggregationFuzzerBase.cpp:366] Executing query plan: 
-- Aggregation[SINGLE [g0] a0 := arbitrary(ROW["c0"]) ORDER BY c0 ASC NULLS LAST] -> g0:ARRAY<BOOLEAN>, a0:TIMESTAMP
  -- Values[1000 rows in 10 vectors] -> c0:TIMESTAMP, g0:ARRAY<BOOLEAN>
I20240327 14:04:16.056669 3310541 Task.cpp:1096] All drivers (1) finished for task test_cursor 1 after running for 30 ms.
I20240327 14:04:16.056702 3310541 Task.cpp:1818] Terminating task test_cursor 1 with state Finished after running for 30 ms.
I0327 14:04:16.683683 3310529 Compression.cpp:628] Initialized zstd compressor with compression level 7
I20240327 14:04:16.692608 3310529 FileSink.cpp:123] closing file:fuzzer.dwrf,  total size: 8.36KB
QueryAssertions.cpp:1129: Failure
Failed
Expected 376, got 376
328 extra rows, 328 missing rows
10 of extra rows:
        null | "3784991-02-06T13:52:39.498000000"
        [] | "10265357-09-13T20:16:11.120000000"
        [null] | "55442863-07-23T06:20:47.805000000"
        [null,null] | "60150461-09-15T15:10:55.522000000"
        [null,null,null,null,null,true,null,null] | "55442863-07-23T06:20:47.805000000"
        [null,null,false] | "181669003-06-08T20:15:29.805000000"
        [null,null,false,true,false,false,false,null,true,true,true,true,null,null,false,null,null,true,false] | "105667027-12-16T05:18:40.658000000"
        [null,null,false,true,false,false,false,true,null,false,false,true,false,false,null,true,true,null,null] | "271030430-06-16T08:10:18.450000000"
        [null,null,false,true,false,true,false,false,false,true] | "11883841-03-28T20:43:26.332000000"
        [null,null,true] | "49077045-08-24T05:22:50.606000000"

10 of missing rows:
        null | "104725217-01-14T04:26:14.072000000"
        [] | "100565777-03-15T01:55:31.427000000"
        [null] | "65877250-09-07T18:37:01.933000000"
        [null,null] | "264175230-11-02T07:20:50.204000000"
        [null,null,null,null,null,true,null,null] | "55442863-07-22T14:20:47.805000000"
        [null,null,false] | "181669003-06-08T04:15:29.805000000"
        [null,null,false,true,false,false,false,null,true,true,true,true,null,null,false,null,null,true,false] | "273329593-02-10T01:01:17.785000000"
        [null,null,false,true,false,false,false,true,null,false,false,true,false,false,null,true,true,null,null] | "271030430-06-15T16:10:18.450000000"
        [null,null,false,true,false,true,false,false,false,true] | "11883841-03-28T04:43:26.332000000"
        [null,null,true] | "278842008-11-07T20:43:18.559000000"

Unexpected results
E20240327 14:04:16.891533 3310529 Exceptions.h:69] Line: AggregationFuzzer.cpp:848, Function:verifySortedAggregation, Expression: assertEqualResults( expectedResult.value(), firstPlan->outputType(), {resultOrError.result}) Velox and reference DB results don't match, Source: RUNTIME, ErrorCode: INVALID_STATE
libc++abi: terminating with uncaught exception of type facebook::velox::VeloxRuntimeError: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Velox and reference DB results don't match
Retriable: False
Expression: assertEqualResults( expectedResult.value(), firstPlan->outputType(), {resultOrError.result})
Function: verifySortedAggregation
File: AggregationFuzzer.cpp
```

And when I run max_by, I got this error, while I think MAP<BOOLEAN,BOOLEAN> should not be passed into the function as a compare key, how can I limit it?
```
./velox_aggregation_fuzzer_test \
                --seed ${RANDOM} \
                --only "max_by" \
                --duration_sec 3600 \
                --logtostderr=1 \
                --minloglevel=0 \
                --repro_persist_path=/tmp/aggregate_fuzzer_repro \
                --enable_sorted_aggregations=true \
                --presto_url=http://127.0.0.1:8080

Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Unknown input types for max_by (SINGLE) aggregation: SMALLINT, MAP<BOOLEAN,BOOLEAN>, BIGINT
Retriable: False
Function: createNArg
File:MinMaxByAggregates.cpp
Line: 1755
```

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp (hunk 15 above)
Comment (f0rest9999) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp]: Fixed and ut covered.

File: velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp [hunk 19]
```diff
@@ -1135,31 +1810,17 @@ exec::AggregateRegistrationResult registerMinMaxBy(
...
-                             .argumentType("V")
-                             .argumentType(compareType)
-                             .argumentType("bigint")
-                             .build());
-  }
+
+  // V, C, bigint -> row(bigint, array(C), array(V)) -> array(V).
+  signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                           .typeVariable("V")
+                           .typeVariable("C")
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp:1847]: @f0rest9999 

> And when I run max_by, I got this error, while I think MAP<BOOLEAN,BOOLEAN> should not be passed into the function as a compare key, how can I limit it?

You may have missed this comment. Once you address it, the error should go away.

**Comment by mbasmanova** [comment]
@f0rest9999 

> That's a good suggestion. Could you open an issue and see if anyone is interested? While I probably haven't had a lot of time to look into how to implement this recently, and I prefer to get this non-simple Function API PR done first to support complex types. Is that all right?

I'm afraid we won't be able to merge this code as is. There is too much copy-paste. The code is too hard to read and reason about. We'll need to simply. I believe the first step would be to convert existing implementation to use Simple Function API. The next step would be to add new logic to it and figure out how to avoid so much copy-paste when doing that. Thanks.

**Comment by f0rest9999** [comment]
> I'm afraid we won't be able to merge this code as is. There is too much copy-paste. The code is too hard to read and reason about. We'll need to simply. I believe the first step would be to convert existing implementation to use Simple Function API. The next step would be to add new logic to it and figure out how to avoid so much copy-paste when doing that. Thanks.

@mbasmanova Thanks for your all review and suggestion, also @kgpai . 
The issue can cc@ me. Rewriting to the simple API will take some time. So it probably not be done quickly by me.
During the time, if someone else has implemented it, I will learn how to do it.In short,  I‘m interested in this issue.

I think the reason it's hard to scale is that templates are used in a lot of places, and the function takes two arguments (various types). The combination of these two reasons made it difficult to extend and reuse code under the old framework.

You can close this PR. Thanks!

**Comment by kgpai** [comment]
> Have you run Aggregation Fuzzer using Presto as source before? In the process of running, I found some problems. @mbasmanova @kgpai
> 
> And when I run max_by, I got this error, while I think MAP<BOOLEAN,BOOLEAN> should not be passed into the function as a compare key, how can I limit it?
> 
> ```
> ./velox_aggregation_fuzzer_test \
>                 --seed ${RANDOM} \
>                 --only "max_by" \
>                 --duration_sec 3600 \
>                 --logtostderr=1 \
>                 --minloglevel=0 \
>                 --repro_persist_path=/tmp/aggregate_fuzzer_repro \
>                 --enable_sorted_aggregations=true \
>                 --presto_url=http://127.0.0.1:8080
> 
> Error Source: RUNTIME
> Error Code: INVALID_STATE
> Reason: Unknown input types for max_by (SINGLE) aggregation: SMALLINT, MAP<BOOLEAN,BOOLEAN>, BIGINT
> Retriable: False
> Function: createNArg
> File:MinMaxByAggregates.cpp
> Line: 1755
> ```
> 
> At present, I don't know much about the data-generate process of Fuzzer, seems like a general problem?

Hi @f0rest9999 this might be related to the your PR change as current min_by, max_by doesnt support complex types for comparison. Nevertheless if you can create an issue and add the stack trace I can investigate more.

**Comment by f0rest9999** [comment]
> Hi @f0rest9999 this might be related to the your PR change as current min_by, max_by doesnt support complex types for comparison. Nevertheless if you can create an issue and add the stack trace I can investigate more.

> You may have missed this comment. Once you address it, the error should go away.
After the modification according to mbasmanova's suggestion, there is no such problem. Thanks for your attention.
